sphinx/ext/intersphinx.py

```diff
@@ -273,22 +274,43 @@ def load_mappings(app: Sphinx) -> None:
         # Duplicate values in different inventories will shadow each
         # other; which one will override which can vary between builds
         # since they are specified using an unordered dict.  To make
-        # it more consistent, we sort the named inventories and then
-        # add the unnamed inventories last.  This means that the
-        # unnamed inventories will shadow the named ones but the named
-        # ones can still be accessed when the name is specified.
-        named_vals = []
-        unnamed_vals = []
+        # it more consistent, we sort the named inventories by their
```

Comment:
I don't think sorting occurs any more with the updated algorithm here?

Reply:
We implicitly sort it thanks to the dict insertion.

Reply:
> We implicitly sort it thanks to the dict insertion

Are you sure about that?  I think that since py37, dictionaries [preserve insertion order](https://docs.python.org/3.7/library/stdtypes.html#dict.values), in which case the order we receive them is dependent on the order in which we load the keys.

Reply:
When you load the cache, you load things that were ordered from a previous build. Now, if you rebuild your project without wiping the build directory, you'll first load those keys and then you iterate over your new mapping. If you encounter an URI that already exists, then you'll directly update the corresponding cache entry, and thus your cache now contains an updated data for that URI. Otherwise, new URIs will be put at the end of the cache.

Now, when you iterate over the cached values, you will first have all items that were present in the old cache (possibly updated because you found the same URI), followed by those that were inserted during your update at the end (those were *not* present in the previous build). So, you'll either get old data or updated data, followed by new data. 

Example:

* Before build 1: cache is empty.
* Config of build 1: `intersphinx_mapping = {'foo': (URL1, None)}`
* After build 1: cache is ``{URL1: ('foo', ..., data1)}``
* Before build 2: cache is ``{URL1: ('foo', ..., data1)}``
* Config of build 2: `intersphinx_mapping = {'foo': (URL2, None)}` (only change the URL).
* During build 2: cache is ``{URL1: ('foo', ..., data1), URL2: ('foo', ..., data2)}``
* When looping over the cache to build the *local* inventory:
  * `value = ('foo', ..., data1)` => `inventory['foo'] <- data1`,
  * `value = ('foo', ..., data2)` => `inventory['foo'] <- data2` (we just replaced).
* During build 2: local inventory is `{'foo': data2}` (no data1 stored in the inventory).
* After build 2: cache is ``{URL1: ('foo', ..., data1), URL2: ('foo', ..., data2)}``.

And thanks to you, I see that we have a growing cache. And maybe this should be changed (like, if you change a lot of your URLs, you'll end up with possibly a very very huge cache; your inventory will still be relatively small during the build but your environment will may store very very huge objects).

Reply:
Ok; I agree that existing `objects.inv` files should contain ordered index content thanks to the `sorted` call in versions to-date.

And it seems that maintaining that invariant (sorted index) is important, since the proposed code here relies on that.

I guess what I'm wary of is that the removal of an explicit sorting operation could make the invariant more fragile.

Reply:
One thing that we still need to address is the fact that old inventories that are *always* replaced at runtime are still stored in the cache. If you change your URL of your project, you'll end up with a cache containing all of the old URLs, even though the intersphinx cache will contain only the inventory data that was fetched from the one in `conf.py`.

I think we must remove those old URLs since they are not used anymore.

Reply:
Do we need to rely on the sort order at all?

(I've now read your [edited GitHub-comment](https://github.com/sphinx-doc/sphinx/pull/11706#discussion_r1347216820), but the code-comment in the branch continues to mention sorting, and I don't think that insertion-sorting is occurring)

Reply:
Actually, now that I think, I *must* remove those old URLs for the following reason:

* Assume that the inventory data is the same for the *different* URLs.
* First build your project with URL1. Your cache will be `{URL1: ("foo", data)}`
* Then change to URL2. Your cache will be `{URL1: ("foo", data), URL2: ("foo", data)}` and your inventory `{"foo": data}`. 
* Then switch back to URL1 but assume that URL1 has now a different data. Your cache will be `{URL1: ("foo", newdata), URL2: ("foo", data)}`. But because I didn't remove URL1 in the previous build, I would end up with my inventory `{"foo": olddata}` since URL2 will be processed after!

So thank you for that.

Reply:
> Do we need to rely on the sort order at all?

Err.. perhaps not. The only case where the local inventories we are constructing (and pickling) have different order occurs when the configuration values change. But maybe I'm missing a case.

Reply:
You're welcome - I think you figured those details out yourself, but I'm glad to help by attempting to write some relatively on-topic things :)

If sorting isn't required then perhaps all that is required here is to update the code-comment.

Reply:
My updated understanding is that the intersphinx cache is stored on the [`BuildEnvironment`](https://github.com/sphinx-doc/sphinx/blob/bb74aec2b6aa1179868d83134013450c9ff9d4d6/sphinx/environment/__init__.py#L138) instance ([`InventoryAdapter.env`](https://github.com/sphinx-doc/sphinx/blob/bb74aec2b6aa1179868d83134013450c9ff9d4d6/sphinx/ext/intersphinx.py#L69)), and that's pickled/[unpickled](https://github.com/sphinx-doc/sphinx/blob/bb74aec2b6aa1179868d83134013450c9ff9d4d6/sphinx/application.py#L314-L324) between builds, explaining how the cache is maintained out-of-process.

(no response required; writing out what I'm learning along the way helps me to figure out and share whether I've understood it.  corrections appreciated if I'm mistaken about any of this!)